### ISSUE TITLE
Error with nikikic php parser

### ISSUE THREAD

**Issue by Aerendir** [rectorphp/rector#3668]
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | v0.7.46              |
| Installed as   | composer dependency                      |

```consoole
PHP Notice:  Undefined offset: 966 in phar:///Path/to/my/lib/vendor/phpstan/phpstan/phpstan.phar/vendor/nikic/php-parser/lib/PhpParser/Internal/TokenStream.php on line 183
Notice: Undefined offset: 966 in phar:///Path/to/my/lib/vendor/phpstan/phpstan/phpstan.phar/vendor/nikic/php-parser/lib/PhpParser/Internal/TokenStream.php on line 183
PHP Notice:  Undefined offset: 884 in phar:///Path/to/my/lib/vendor/phpstan/phpstan/phpstan.phar/vendor/nikic/php-parser/lib/PhpParser/Internal/TokenStream.php on line 199
Notice: Undefined offset: 884 in phar:///Path/to/my/lib/vendor/phpstan/phpstan/phpstan.phar/vendor/nikic/php-parser/lib/PhpParser/Internal/TokenStream.php on line 199
PHP Notice:  Undefined offset: 889 in phar:///Path/to/my/lib/vendor/phpstan/phpstan/phpstan.phar/vendor/nikic/php-parser/lib/PhpParser/Internal/TokenStream.php on line 183
Notice: Undefined offset: 889 in phar:///Path/to/my/lib/vendor/phpstan/phpstan/phpstan.phar/vendor/nikic/php-parser/lib/PhpParser/Internal/TokenStream.php on line 183
```

## Minimal PHP Code Causing Issue

Impossible to reproduce: the demo removes the results file:

https://getrector.org/demo/86bb8198-933b-4651-b5a9-58f79fcab5e3

I've tried to go back with versions of `rector/rector`, `nikic/php-parser` and `phpstan/phpstan` but without success.

The last configuration of Composer I tried was this:

```json
"require-dev": {
        "rector/rector": "^0.6 | ^0.7"
    },
    "conflict": {
        "nikic/php-parser": ">=4.5.0",
        "phpstan/phpstan": ">=0.12.32"
    }
```

But at each update I got other errors, so I think it is better to try to fix the current version of Rector.

**Comment by TomasVotruba** [comment]
How can we reproduce?

**Comment by Aerendir** [comment]
@TomasVotruba , ok let's start small: which sets do you think I can remove?

For example: all the sets about Symfony: it is sufficient to simply use one for SF5? I need to support more versions of Symfony...

**Comment by TomasVotruba** [comment]
> ok let's start small: which sets do you think I can remove

I remove all of them by half reduction method, untill I find the one.

**Comment by Aerendir** [comment]
@TomasVotruba , I was speaking about this phrase:

> usually only 1 rule is needed, not 40 sets.
> It will speed up the fix by weeks

But reading again it, I now understand what you said: I thought you were suggesting me to remove the sets from MY config, not the one provided in the demo.

Anyway, I found half an hour to continue a bit of debugging: the set that causes the  error is the `psr-4` one and I suspect the cause is the fact that my project doesn't strictly follows it.

The rule that causes the problem is this:

`Rector\PSR4\Rector\Namespace_\NormalizeNamespaceByPSR4ComposerAutoloadRector`

I have a folder structure like this:

```
src/
  ClassOne.php
  ClassTwo.php
  ClassThree.php
```

And their namspace is something like `My\App\`, mapped in `composer.json` this way:

```
    "autoload": {
        "psr-4": {
            "My\\App\\": "src"
        }
    },
```

Maybe this is the cause.

Is this of any help to you?

**Comment by Aerendir** [comment]
Errata corridge: the class that causes the error is `Rector\PSR4\Rector\MultipleClassFileToPsr4ClassesRector`.

**Comment by TomasVotruba** [comment]
Good :+1: Now we need failing test fixture to be added in `MultipleClassFileToPsr4ClassesRectorTest`

**Comment by TomasVotruba** [comment]
How did you solve it? It might be useful to others with similar issues in the future

**Comment by Aerendir** [comment]
@TomasVotruba , I didn't: I simply moved all dev deps on bamarni/composer-bin-plugin.

Now I have non more conflicts as each tool (phan, phpstan, rector, phpunit, psalm, etc...) has each its own `composer.json` and its own dependencies.

And I also disabled the rule as in this moment my goal is to get green builds on each component, bundle and app I manage (over 25 packages 😰).

**Comment by TomasVotruba** [comment]
> Now I have non more conflicts as each tool (phan, phpstan, rector, phpunit, psalm, etc...) has each its own composer.json and its own dependencies.

For legacy project, we use stanalone builds in Gitlab/Github CI for this with their own image. 
It's pretty fast and easy to setup, no conflicts.

**Comment by Aerendir** [comment]
@TomasVotruba , yes... lets me explain...

Mine are not legacy projects but maintained ones and upgraded, too.

Having the dev deps in the main composer, causes the dependency hell that we all know.

Using the Phars is one of the possible solutions.

I managed them through Phive, to have a bit of centralized control.

But now, I migrated all the projects on GitHub and GitHub has Dependabot that takes care of the update of the libraries.

But it doesn't support Phars.

instead, it supports multiple `composer.json`.

So, basically, using bamarni's plugin, I can:

1. Avoid dependency hell
2. Have latest versions of all dependencies I need, without worrying about conflicts
3. Have Dependabot update the dependencies for me
4. Ensure in CI that each update/upgrade doesn't break anything

For example, Dependabot has just alerted me that there is a new version of Phan but the build broke.

Now I can check which is the problem, update and use ASAP the lateste version of Phan.

And this is valid for any dependency.

It's another world!
